spec/ordinary-and-exotic-object-behaviours.html

```diff
@@ -0,0 +1,97 @@
+        <emu-alg>
+          1. Assert: IsPropertyKey(_P_) is *true*.
+          1. If Type(_P_) is a Symbol, throw a *TypeError* exception.
+          1. Return ! RecordGetOwnProperty(_R_, _P_).
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-record-exotic-objects-defineownproperty-p-desc">
+        <h1>[[DefineOwnProperty]] ( _P_, _Desc_ )</h1>
+        <p>When the [[DefineOwnProperty]] internal method of a Record exotic object _R_ is called with property key _P_, and Property Descriptor _Desc_ the following steps are taken:</p>
+        <emu-alg>
+          1. Assert: IsPropertyKey(_P_) is *true*.
+          1. Let _current_ be ? _R_.[[GetOwnProperty]](_P_).
+          1. If _current_ is *undefined* return *false*.
+          1. If IsAccessorDescriptor(_Desc_) is *true*, return *false*.
+          1. If _Desc_.[[Writable]] is present and has value *false*, return *false*.
+          1. If _Desc_.[[Enumerable]] is present and has value *false*, return *false*.
+          1. If _Desc_.[[Configurable]] is present and has value *true*, return *false*.
+          1. If _Desc_.[[Value]] is present, return SameValue(_Desc_.[[Value]], _current_.[[Value]]).
+          1. Return *false*.
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-record-exotic-objects-ownpropertykeys">
+        <h1>[[OwnPropertyKeys]] ( )</h1>
+        <p>When the [[OwnPropertyKeys]] internal method of a Record exotic object _R_ is called, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _keys_ be a new empty List.
+          1. Let _rec_ be _R_.[[RecordData]].
+          1. Assert: Type(_rec_) is Record.
+          1. Let _len_ be the length of _rec_.
+          1. For each pair _p_ of _rec_.[[Fields]], do
+            1. Let _key_ be _p_.[[Key]]
+            1. Assert: Type(_key_) is String
+            1. Add _key_ to _keys_
+          1. Return _keys_
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-recordcreate" aoid="RecordCreate">
+        <h1>RecordCreate ( _value_ )</h1>
+        <p>This abstract operation RecordCreate takes a _value_ argument. It is used to specify the creation of new Record exotic objects. It performs the following steps when called:</p>
+        <emu-alg>
+            1. Assert: Type(_value_) is Record.
+            1. Let _R_ be ! MakeBasicObject(&laquo; [[RecordData]] &raquo;).
+            1. Set _R_.[[RecordData]] to _value_
+            1. Set _R_.[[GetOwnProperty]] as specified in <emu-xref href="#sec-record-exotic-objects-getownproperty-p"></emu-xref>.
+            1. Set _R_.[[DefineOwnProperty]] as specified in <emu-xref href="#sec-record-exotic-objects-defineownproperty-p-desc"></emu-xref>.
+            1. Set _R_.[[OwnPropertyKeys]] as specified in <emu-xref href="#sec-record-exotic-objects-ownpropertykeys"></emu-xref>.
```

Comment:
Some things missing here off-hand:
- You haven't done PreventExtensions and need to
- The prototype needs to be null somehow (either by declaring a [[Prototype]] slot and setting it to null, or by making SetPrototypeOf/GetPrototypeOf methods that act like it's null)

tbh I was picturing that all the methods would be overridden from scratch, like in [module namespace exotic objects](https://tc39.es/ecma262/#sec-module-namespace-exotic-objects), rather than falling back to ordinary object behavior like strings do. You're not using any of the object internal slots, though some things do "work out" with this phrasing anyway (such as [[HasProperty]]).

Reply:
this is an outdated diff: I am in the process of getting things from module ns exotic object, still not done though